zero_noise_extrapolation_cnot.py

```diff
@@ -0,0 +1,570 @@
+from qiskit import QuantumCircuit, execute, Aer, transpile
+
+from qiskit.transpiler import PassManager, PassManagerConfig, CouplingMap
+from qiskit.transpiler.passes import Unroller, Optimize1qGates
+from qiskit.transpiler.preset_passmanagers.level3 import level_3_pass_manager
+
+from numpy import asarray, ndarray, shape, zeros, empty, average, transpose, dot
+from numpy.linalg import solve
+
+import random
+
+
+"""
+-- ZERO NOISE EXTRAPOLATION for CNOT-gates --
+
+This implemention does quantum error mitigation using the method of zero noise extrapolation, by amplifying noise in
+a quantum circuit by a set of noise amplification factors, then using Richardson extrapolation to extrapolate the
+expectation value to the zero-noise limit.
+
+The noise amplified and mitigated is specifically noise in CNOT-gates. The noise is amplified by n amplification
+factors 1, 3, 5, ..., 2n + 1, where 1 means the bare circuit without noise amplification, and 3 means every
+CNOT gate is extended as CNOT*CNOT*CNOT.
+
+As CNOT*CNOT = Id, the identity, in the noise-free case the amplified CNOT have the same action as a single CNOT, 
+but ~3 times the noise.
+
+"""
+
+
+class ZeroNoiseExtrapolation:
+
+    def __init__(self, qc: QuantumCircuit, exp_val_func, backend=None, noise_model=None,
+                 n_amp_factors: int = 3, pauli_twirl: bool = False, shots: int = 8192,
+                 pass_manager: PassManager = None):
+        """
+        :param qc: The circuit to be mitigated
+        :param exp_val_func: A function which computes the observed expectation value of some operator as a function of
+                             the measurement counts from the circuit qc
+        :param backend: The backend on which to execute the circuit
+        :param noise_model: Optinal custom noise model for execution on a simulator backend
+        :param n_amp_factors: Number of amplification factors to be used. For n amplification factors, the
+                              amplification factors will be 1,3,5,...,2n + 1
+        :param pauli_twirl: Do pauli twirl True / False
+        :param shots: Number of shots of the circuit to execute
+        :param pass_manager: Optional custom pass manager to use when transpiling the circuit
+        """
+
+        if backend == None:
+            self.backend = Aer.get_backend("qasm_simulator")
+            self.is_simulator = True
+        else:
+            self.backend = backend
+            self.is_simulator = backend.configuration().simulator
+
+        # Do an initial heavy optimization of the input circuit
```

Comment:
This is optional, right? If we provide an empty pass_manager, nothing will be done.